sw/device/tests/sim_dv/otp_ctrl_lc_signals_test.c

```diff
@@ -0,0 +1,378 @@
+ *
+ * The test can be run in different life cycle states and figure out what to
+ * check based on the state exposed in the life cycle controller.
+ *
+ * PROD/DEV/RMA:
+ * 1)  Provision non-constant creator/owner secrets to flash
+ * 2)  Program SECRET2 partition and read it back
+ * 3)  Check that the key manager advance errors out due to all-zero root key
+ * 4)  Reset the chip
+ * 5)  Check that the SECRET2 partition can still be read
+ * 6)  Check that the key manager advance errors out due to all-zero root key
+ * 7)  Lock the SECRET2 partition
+ * 8)  Reset the chip
+ * 9)  Check that the SECRET2 partition is not accessible anymore
+ * 10) Check that the key manager can generate a SW key without errors
+ *
+ * All other life cycle states:
+ * 1) Check that key manager cannot be initialized because it is disabled
+ * 2) Check the SECRET2 partition is not accessible
+ *
+ */
+bool test_main(void) {
+  bool is_computed;
+  dif_lc_ctrl_state_t state;
+  dif_rstmgr_reset_info_bitfield_t rst_info;
+  init_peripherals();
+
+  rst_info = rstmgr_testutils_reason_get();
+  rstmgr_testutils_reason_clear();
+
+  CHECK_DIF_OK(dif_lc_ctrl_get_state(&lc, &state));
+  lc_ctrl_testutils_print_lc_state(&state);
+
+  switch (state) {
+    case kDifLcCtrlStateDev:
+    case kDifLcCtrlStateProd:
+    case kDifLcCtrlStateProdEnd:
+    case kDifLcCtrlStateRma:
+      if (rst_info & kDifRstmgrResetInfoPor) {
+        LOG_INFO("First access test iteration...");
+        // Make sure the secrets in flash are non-zero.
+        keymgr_testutils_init_flash(&flash, &kCreatorSecret, &kOwnerSecret);
+        // Program the SECRET2 partition and perform read back test.
+        run_otp_access_tests(kWriteReadMode, kExpectPassed);
+        // We expect the root key to be invalid at this point.
+        keymgr_check_root_key_is_invalid();
+        reset_chip();
+
+      } else if (rst_info == kDifRstmgrResetInfoSw) {
+        CHECK_DIF_OK(dif_otp_ctrl_is_digest_computed(
+            &otp, kDifOtpCtrlPartitionSecret2, &is_computed));
+
+        if (!is_computed) {
+          LOG_INFO("Second access test iteration...");
+          // SECRET2 partition should still be readable.
+          run_otp_access_tests(kReadMode, kExpectPassed);
+          // We expect the root key to still be invalid at this point, since
+          // SECRET2 has not been locked yet.
+          keymgr_check_root_key_is_invalid();
+          // Lock the SECRET2 partition.
+          otp_ctrl_testutils_lock_partition(&otp, kDifOtpCtrlPartitionSecret2,
+                                            0);
+          reset_chip();
+        } else {
+          LOG_INFO("Third access test iteration...");
+          // All accesses are disallowed.
+          run_otp_access_tests(kWriteReadMode, kExpectFailed);
+          // At this point we expect that the key manager can generate keys
+          // without errors.
+          keymgr_check_can_generate_key();
```

Comment:
nit: Add `return true` here, and at the end of the function update to `return false`, just to make sure there are no weird corner cases uncovered by the test.

Reply:
good point - done!